File: datacenter-proxies/Nodejs/scraper.js
const client = require('./client');
const settings = require('./settings');
const filesystem = require('./filesystem');

module.exports = {
  scrape: async (position, proxy, url) => {
    let response = null;
    let retry = 0;
    do {
      [response, err] = await client.fetchPage(proxy, url);
      if (response !== null && response.status !== 200) {
        await filesystem.writeErrorToFile(`${url} - Response code ${response.status}`)
      } else if (err !== null) {
        await filesystem.writeErrorToFile(`${url} - Response error ${err.message}`)
        filesystem.writeErrorToStdout(`${url} failed with error ${err.message}`)
      }

      if (response !== null && response.status === 200) {
        await filesystem.writeSuccessToFile(position, response.data)
        break
      }

      retry += 1;
    } while (retry < settings.RetriesNum);
  }
};

File: datacenter-proxies/Ruby/scraper.rb
require_relative './client'
require_relative './settings'
require_relative './filesystem'

class Scraper
  def initialize(client)
    @client = client
  end

  def scrape(position, proxy, url)
    attempt = 0

    loop do
      response = @client.fetch_page(proxy, url)
      if response != nil && response.code != "200"
        write_error_to_file("#{url} - Response code #{response.code}")
        write_error_to_stdout("#{url} - Response code #{response.code}")
      end

      if response != nil && response.code == "200"
        write_success_to_file(position, response.body)
        break
      end

      attempt += 1
      if attempt >= get_setting(:RetriesNum)
        break
      end
    end
  end
end
